File: client/src/types/chatroom.ts
import { Coordinates } from "./location";

export type Member = {
  id: number;
  username: string;
  avatar_url: string;
};

export type ChatRoom = {
  id: string;
  room_name?: string;
  members_count?: number;
  coordinates?: Coordinates;
  room_type?: string;
  current_user_has_joined?: boolean;
  chatroom_location_type?: string;
  unread_messages_count?: number;
  members?: Member[];
  last_message?: {
    content: string;
    created_at: string;
  };
  updated_at?: string;
  messages_count?: number;
  chat_room_members?: Member[];
};

export type ChatShowProps = {
  selectedChatroom: ChatRoom;
  setSelectedChatroom: (chatroom: ChatRoom) => void;
  isShow?: boolean;
  handleCloseDetail: () => void;
};

export type ChatRoomUpdatePayload = {
  chat_room_id: string;
  room_name?: string;
  last_message: {
    content: string;
    created_at: string;
  };
  updated_at: string;
  sender_name: string;
  unread_messages_count?: number;
  members_count?: number;
};

export type ChatRoomMapResponse = {
  id: number;
  room_name: string;
  coordinates: Coordinates;
  members_count: number;
};

export type ChatRoomMap = {
  id: number;
  room_name: string;
  coordinates: Coordinates;
  members_count: number;
};

export type AutoJoinByLocationRespone = {
  status: number;
  message: string;
};

File: server/app/api/entities/chat_rooms/chat_room.rb
module Entities
  module ChatRooms
    class ChatRoom < Grape::Entity
      expose :id
      expose :room_name
      expose :room_type
      expose :chatroom_location_type
      expose :coordinates do |chat_room, _options|
        {
          lat: chat_room.latitude,
          lng: chat_room.longitude
        }
      end

      expose :members_count
      expose :current_user_has_joined do |chat_room, options|
        options[:current_user_has_joined] || false
      end

      expose :unread_messages_count, if: ->(_, options) { options[:show_unread_count] } do |chat_room, options|
        unread_counts = options[:unread_counts] || {}
        unread_counts[chat_room.id] || 0
      end

      expose :last_message, if: ->(_, options) { options[:last_messages].present? } do |chat_room, options|
        last_message = options[:last_messages][chat_room.id]
        if last_message
          {
            content: last_message.content,
            created_at: last_message.created_at,
          }
        end
      end

      expose :chat_room_members, if: ->(_, options) { options[:chat_room_members].present? } do |chat_room, options|
        members = options[:chat_room_members][chat_room.id.to_i] || []
        members.map do |member|
          {
            id: member[:id],
            username: member[:username],
          }
        end
      end

      expose :members, if: ->(_, options) { options[:show_members] } do |chat_room, options|
        members = options[:members] || chat_room.users.select(:id, :username)
        members.map do |member|
          {
            id: member.id,
            username: member.username
          }
        end
      end

      expose :updated_at do |room, options|
        updated_ats = options[:updated_ats]
        if updated_ats && updated_ats[room.id]
          updated_ats[room.id].iso8601
        else
          room.updated_at.iso8601
        end
      end
    end
  end
end
